### ISSUE TITLE
Extended identifiers-why and where, exactly?

### ISSUE THREAD

**Issue by fecundf** [Raku/problem-solving#224]
Extended identifiers have vague boundaries as to where they are legal.

These thoughts come from the docs at https://docs.raku.org/syntax/identifiers#Extended_identifiers

It begins well with examples of extended module names and operator names, which establishes the _intent_ for extended identifiers.

It then gets murky by showing plain scalar variables with extended names.

And a little murkier as it explains "Where an extended identifier comprises two or more colon pairs, their order is generally significant ... An exception to this rule is module versioning" - why is that - should it be that way?

And it does not state if extended identifiers are meant to work everywhere that normal identifiers do. Are they?

In short the issues are
## Nowhere explaining if extended identifiers are meant to have any limits in their scope

For the scope, see mailing list discussion "Extended identifiers in named attributes" https://www.nntp.perl.org/group/perl.perl6.users/2020/08/msg9096.html where they don't work in signature named args, is that by design or is it a compiler bug?

**Comment by fecundf** [comment]
As for the "Inconsistency between order of identifiers in module versioning vs other identifiers"
It is useful for module versions to have order-independent components–can anyone's code also get that independent-ness?

**Comment by MARTIMM** [comment]
P.s. I made a writing error in terms; I meant argument instead of attribute. I wanted to name an argument using an extended identifier format.

It came about in a project of mine **XML::Actions**. Element tags found in the XML are delivered to user-defined methods. To differentiate between the start and end of an element notation, the user can define methods `element:start` and `element:end`. This works fine. The XML attributes were given as named arguments. One of the users mentioned something involving XML namespaces. For method names there is still no problem, I can define `ns:element:start`. But a named argument like `:$ns:attribute` is generating errors. In that case I can only use a slurpy hash `*%attributes` and process it like `%attributes<ns:attribute>`.

**Comment by fecundf** [comment]
Your example `ns:element:start` as a method name shows that position-significance is useful- if the schema has tags `<start>...</start>`, `<end>...</end>` - ns:start:end and ns:end:start mean different things!

**Comment by fecundf** [comment]
.. uh-oh _"Starting with Raku language version 6.d, colon pairs with sym as the key (e.g. :sym<foo>) are reserved for possible future use"_ that means, **XML::Actions** won't work with a tag `<sym>` in a namespace, requiring a method `ns:sym` - I am opening a new issue on that. https://github.com/Raku/problem-solving/issues/225

Continuing on the _"order of identifiers in module versioning vs other identifiers"_ thought- Module loaders have a way of doing something similar to dispatch, given an extended identifier, and a library of modules. Which explains the note in the docs "Where an extended identifier comprises two or more colon pairs, their order is generally significant ... An exception to this rule is module versioning"

1. Is module version independence a feature of the syntax, with `use a:b:c` parsing into an independent-position ident and `sub a:b:c` parsing into a significant-position ident? (I think that is NOT the case, interesting if it is!)
2. Is module version independence a feature of `use`, which breaks apart the identifier for its lookup?

If the answer is 2, how about changing the wording.

Where an extended identifier comprises two or more colon pairs, their order is _(remove "generally")_ significant ...
With _module versioning_ the importer gives all tags equal consideration, so these identifiers name the same module:

That would satisfy my concern on the "order of identifiers in module versioning", mostly.
I would still be curious how that happens and if it is useful to expose the mechanism, just a little.
